odoo/addons/base/tests/test_ir_mail_server_smtpd.py

```diff
@@ -0,0 +1,651 @@
+import logging
+import shutil
+import smtplib
+import ssl
+import subprocess
+import tempfile
+import unittest
+import warnings
+from base64 import b64encode
+from collections import namedtuple
+from pathlib import Path
+from socket import getaddrinfo
+from unittest.mock import patch
+
+from odoo.exceptions import UserError
+from odoo.tools import file_path, mute_logger
+from .common import TransactionCaseWithUserDemo
+
+try:
+    import aiosmtpd
+    import aiosmtpd.controller
+    import aiosmtpd.smtp
+    import aiosmtpd.handlers
+except ImportError:
+    aiosmtpd = None
+
+
+PASSWORD = 'secretpassword'
+PORT = 8025
+_openssl = shutil.which('openssl')
+_logger = logging.getLogger(__name__)
+
+
+def _smtp_authenticate(server, session, enveloppe, mechanism, data):
+    """ Callback method used by aiosmtpd to validate a login/password pair. """
+    result = aiosmtpd.smtp.AuthResult(success=data.password == PASSWORD.encode())
+    _logger.info("AUTH %s", "successfull" if result.success else "failed")
+    return result
+
+
+def _generate_certificates(certdir):
+    """Generate and sign a few RSA key pairs inside certdir.
+
+    It invokes the openssl command line to create a new Certification
+    Authority and then use that CA to generate extra keys and
+    certificates for a user, a server, and any purpose.
+
+    It also creates a self-signed certificate, to test that invalid
+    certificates are correctly rejected.
+
+    It returns 4 RSA: CA, web server, mail server, self-signed. Each RSA
+    is a (named)tuple of 2 elements: key and cert.
+    """
+    # Create a fake certificate authority and use it to generate
+    # a few signed certificates.
+    shutil.copy(file_path('base/tests/ssl/openssl.conf'), certdir)
+    shutil.copy(file_path('base/tests/ssl/gencert.sh'), certdir)
+    subprocess.run(
+        [shutil.which('sh'), '-e', certdir/'gencert.sh'],
+        check=True,
+        cwd=certdir,
+        env={'OUTDIR': certdir},
+        capture_output=not _logger.isEnabledFor(logging.DEBUG),  # show output in DEBUG
+    )
+
+    # Collect files and return them
+    RSA = namedtuple('RSA', ('key', 'cert'))
```

Comment:
I won't call it RSA, maybe Certificate ?
(you can technically have certificate with ECC and it will work)